src/python/pants/engine/native.py

```diff
@@ -283,6 +284,25 @@ def extern_get_type_for(self, context_handle, val):
     type_id = c.to_id(type(obj))
     return TypeId(type_id)
 
+  @_extern_decl('Handle', ['ExternContext*', 'TypeId'])
+  def extern_get_handle_from_type_id(self, context_handle, ty):
+    c = self._ffi.from_handle(context_handle)
+    return c.to_value(ty)
+
+  @_extern_decl('TypeId', ['ExternContext*', 'TypeId'])
+  def extern_get_union_for(self, context_handle, type_id):
```

Comment:
I think that my comment mislead you here: sorry. Rather than getting the union for a type (which is ambiguous: a type might be in multiple unions), instead it would be easiest to check whether a particular type _is_ a union, and then use that on the "middle" argument from the `Get` to decide whether you should render the union specific error message.

A few potential ways to do that:
1. add a boolean "is_union" extern, that checks whether the union has the magic property added by the `def union` method.
2. add a version of `extern_project_*` that is fallible, and that might result in `None` if the field doesn't exist (ie, equivalent to `getattr(x, y, None)`), and then use that to access the magic "I'm a union" field.

----

Both of these also have the advantage of not needing the BuildConfiguration in here, which should simplify things. Sorry!